enhancements/results-forwarding.md

```diff
@@ -0,0 +1,504 @@
+compliance scans we could simply compress the gathered objects the
+`api-resource-collector` downloads and send that too.
+
+While evidence is important to consider, we're going to dedicated a separate
+enhancement to that problem. This particular enhancement is focused on
+forwarding results.
+
+### Goals
+
+* Provide a generic way to implement compliance result forwarding for
+  Compliance Operator. This should be flexible enough to accommodate the
+  current implementation (CRDs and PVCs), while allowing for custom forwarding
+  implementations.
+
+* Define and implement a stable API for forwarder implementations.
+
+* Implement an interface for forwarding result CRDs
+
+* Implement an interface for forwarding remediation CRDs
+
+* Implement a switch to disable in-cluster storage if forwarding is enabled and valid
+
+### Non-Goals
+
+* Deprecate the current CRD/PVC approach: There is still a use-case for this.
+
+* Expand the "evidence" that's currently gathered by the Compliance Operator. This is
+  needed but should be done as part of a separate initiative.
+
+* Implement an archiving solution. That's being discussed as a separate
+[enhancement](https://github.com/ComplianceAsCode/compliance-operator/pull/66).
+
+## Proposal
+
+The Compliance Operator will forward compliance results and remediation
+recommendations if the supplied forwarding endpoint is valid. The Compliance
+Operator will also continue to use CRDs and PVCs to store results,
+remediations, and evidence as it does today.
+
+A user may explicitly disable in-cluster storage if forwarding is enabled. The
+Compliance Operator will not support disabling PVC storage without a valid
+forwarding endpoint.
+
+This will require changing the **aggregator** to always forward via a gRPC API
+to a configured implementation. In this case, the **aggregator** will act as a
+gRPC client. The **resultserver** then becomes a gRPC server, subject to the
+evidence forwarding provider and will be renamed **evidence-persistor** and
+responsible for writing results.
+
+Forwarder implementations would be configured via the `ScanSetting` and they'd be called "providers".
+For backwards compatibility, we'd have defaults that would point to the current mode of operation
+of the Compliance Operator. e.g. a `pvc` provider for evidence storage.
+
+For more detail on the proposed API, see the *API Extensions* section.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters, I'd like to run
+  compliance scans on my fleet in such a way that uses the minimal amount of
+  resources.
+
+* As a Site Reliability Engineer managing multiple clusters, I'd like to have
+  one place to view the compliance stance of my whole fleet.
+
+### API Extensions
+
+Given the scope of the change, we'd upgrade the `ScanSetting` version to
+`v1alpha2`. The following sections describe the additional values needed by the
+`ScanSetting` CRD.
+
+#### `resultForwarding.provider`
+
+Required: True
```

Comment:
Possibly a dumb question, but what does 'Required' mean here? "Does not have a default and must be set"?
Shouldn't `grpc` to the local crd-creating service be the default?

Reply:
So of course the downside of having the grpc service exposed over local sockets only is that we don't test the network forwarding and would have more configurations to test..

With the grpc-over-network approach, how would the default CO deployment look like then? I was imagining we'd have one more deployment with the grpc server, a service and then the grpc client would use `http://service` as the endpoint (and since we'd know the service name, everything could just use defaults..)

Reply:
@mrogers950 do you have an opinion?

Reply:
I'd say always requiring the gRPC traffic to go through the network would help in keeping things consistent. e.g. you wouldn't need a code-path to handle creating a unix socket and having a separate goroutine in the aggregator to handle this.

Reply:
I also agree that we should choose a sane default that's backwards compatible, that is what users expect and would help in migration.

Reply:
It sounds like this should still be required, then?

Should we also have a field for marking a default value?

Reply:
On the topic of the local gRPC traffic, since this is only results, I think that making a `local` provider and all would be overkill (requires the aggregator to break up into a client/server).  For the evidence proposal that type of pattern makes more sense for what we want to accomplish there, since that's where our existing component client+servers are involved.

Reply:
hmm, I guess your proposal would also be safer (more stable) in the sense that the aggregator wouldn't change that drastically from a self-contained binary to a client/server.

Just FWIW, the reason I proposed the always-forwarder was just that it felt architecturally (even if local) much cleaner where the aggregator would be just a forwarder to a handler, because we're have to implement the forwarding either way.

Reply:
I'm not opposed to moving towards that if it makes sense for the rest of the forwarding features. We could still build the server implementation but keep it as a test component to run the aggregator client against.

Reply:
that sounds like a good compromise. We need a testbed anyway and having a test server that produces the CRDs would also make it easier to do e2e testing.

Good idea!